File: src/providers/zenvia/scheduler.ts
import { randomUUID } from "node:crypto";
import { registerJobHandler, scheduleJob } from "../../shared/scheduler/scheduler.js";
import { deliverWebhook } from "../../shared/webhooks/deliver.js";
import { findActiveSubscriptionsForChannel, getMessage, updateMessageStatus } from "./state.js";

const STATUS_DELAY_MS = Number(process.env.ZENVIA_STATUS_DELAY_MS ?? 2000);

const NEXT_STATUS: Record<string, string | undefined> = {
  ACCEPTED: "SENT",
  SENT: "DELIVERED",
};

interface AdvanceJobPayload {
  messageId: string;
}

export function scheduleStatusAdvance(messageId: string): void {
  scheduleJob("zenvia:advance", { messageId } satisfies AdvanceJobPayload, new Date(Date.now() + STATUS_DELAY_MS));
}

async function handleAdvance(payload: unknown): Promise<void> {
  const { messageId } = payload as AdvanceJobPayload;
  const message = getMessage(messageId);
  if (!message) return;

  const nextStatus = NEXT_STATUS[message.status];
  if (!nextStatus) return;

  updateMessageStatus(messageId, nextStatus);

  const subscriptions = findActiveSubscriptionsForChannel(message.channel);
  for (const subscription of subscriptions) {
    if (subscription.criteriaDirection && subscription.criteriaDirection !== "ALL" && subscription.criteriaDirection !== message.direction) {
      continue;
    }
    await deliverWebhook({
      provider: "zenvia",
      resourceType: "message",
      resourceId: message.id,
      url: subscription.webhookUrl,
      headers: subscription.webhookHeaders ?? undefined,
      payload: {
        id: randomUUID(),
        timestamp: new Date().toISOString(),
        subscriptionId: subscription.id,
        type: "MESSAGE_STATUS",
        channel: message.channel,
        message: {
          id: message.id,
          externalId: message.externalId ?? undefined,
          direction: message.direction,
          from: message.from,
          to: message.to,
        },
        messageStatus: {
          code: nextStatus,
          timestamp: new Date().toISOString(),
          channel: message.channel,
          direction: message.direction,
        },
      },
    });
  }

  if (NEXT_STATUS[nextStatus]) {
    scheduleStatusAdvance(messageId);
  }
}

export function registerZenviaScheduler(): void {
  registerJobHandler("zenvia:advance", handleAdvance);
}

File: internal/providers/zenvia/scheduler.go
package zenvia

import (
	"context"
	"database/sql"
	"encoding/json"
	"time"

	"github.com/google/uuid"

	"github.com/golgimed/mimic/internal/shared/admin"
	"github.com/golgimed/mimic/internal/shared/scheduler"
	"github.com/golgimed/mimic/internal/shared/webhooks"
)

const advanceJobKind = "zenvia:advance"

var nextStatus = map[string]string{
	"ACCEPTED": "SENT",
	"SENT":     "DELIVERED",
}

type advanceJobPayload struct {
	MessageID string `json:"messageId"`
}

// RegisterScheduler wires the zenvia:advance job handler and returns a
// ScheduleAdvance closure providers/handlers use to enqueue the next hop
// (ACCEPTED -> SENT -> DELIVERED) for a freshly created message.
func RegisterScheduler(sched *scheduler.Scheduler, db *sql.DB, store *Store, faultStore *admin.Store, statusDelay time.Duration) func(messageID string) error {
	scheduleAdvance := func(messageID string) error {
		return sched.ScheduleJob(advanceJobKind, advanceJobPayload{MessageID: messageID}, time.Now().Add(statusDelay))
	}

	sched.RegisterJobHandler(advanceJobKind, func(ctx context.Context, raw json.RawMessage) error {
		var payload advanceJobPayload
		if err := json.Unmarshal(raw, &payload); err != nil {
			return err
		}

		message, err := store.GetMessage(payload.MessageID)
		if err != nil || message == nil {
			return err
		}

		next, ok := nextStatus[message.Status]
		if !ok {
			return nil
		}

		if err := store.UpdateMessageStatus(payload.MessageID, next); err != nil {
			return err
		}

		subscriptions, err := store.FindActiveSubscriptionsForChannel(message.Channel)
		if err != nil {
			return err
		}

		for _, sub := range subscriptions {
			if sub.CriteriaDirection != nil && *sub.CriteriaDirection != "ALL" && *sub.CriteriaDirection != message.Direction {
				continue
			}

			eventPayload := map[string]any{
				"id":             uuid.NewString(),
				"timestamp":      time.Now().UTC().Format(time.RFC3339Nano),
				"subscriptionId": sub.ID,
				"type":           "MESSAGE_STATUS",
				"channel":        message.Channel,
				"message": map[string]any{
					"id":         message.ID,
					"externalId": message.ExternalID,
					"direction":  message.Direction,
					"from":       message.From,
					"to":         message.To,
				},
				"messageStatus": map[string]any{
					"code":      next,
					"timestamp": time.Now().UTC().Format(time.RFC3339Nano),
					"channel":   message.Channel,
					"direction": message.Direction,
				},
			}

			webhookFault, err := faultStore.ConsumeMatchingFault(Name, "webhook")
			if err != nil {
				return err
			}

			if webhookFault != nil && webhookFault.FaultKind == admin.FaultWebhookDropped {
				if err := webhooks.LogDropped(db, webhooks.DeliverInput{
					Provider: Name, ResourceType: "message", ResourceID: message.ID,
					URL: sub.WebhookURL, Payload: eventPayload,
				}); err != nil {
					return err
				}
				continue
			}

			if webhookFault != nil && webhookFault.FaultKind == admin.FaultWebhookInvalid {
				delete(eventPayload, "messageStatus")
				eventPayload["malformed"] = true
			}

			if err := webhooks.Deliver(ctx, db, webhooks.DeliverInput{
				Provider: Name, ResourceType: "message", ResourceID: message.ID,
				URL: sub.WebhookURL, Headers: sub.WebhookHeaders, Payload: eventPayload,
			}); err != nil {
				return err
			}
		}

		if _, hasNext := nextStatus[next]; hasNext {
			return scheduleAdvance(payload.MessageID)
		}
		return nil
	})

	return scheduleAdvance
}
